### PR TITLE
Quadratures

### LINKED ISSUE #2511: quadrature mark
**Is your feature request related to a problem? Please describe.**
A way to see where quadrature happens (see also #789)

**Describe the solution you'd like**
Ideas:
- a new line "quadrature line" (actually a cone centered on the Sun)
- a quadrature mark on the ecliptic

**Describe alternatives you've considered**
Search for elongation at 90/270 or use astrocalc

**Additional context**

### PR DESCRIPTION
**Pull request by alex-w**
### Description
Added visualisation the line of quadratures and compute the approximated time of quadratures for outer planets into AstroCalc/Phenomena tool.

Fixes #2511, #789

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by worachate001** [comment]
It looks like there is a problem in AstroCalc. My calculation indicates that the upcoming Mars's western quadrature will occur on 2022 August 27 (05:28 UT). But Stellarium gives this incorrect result(s).
Mars-quadrature

**Comment by alex-w** [comment]
@worachate001 could you share your location name to testing?

**Comment by worachate001** [comment]
> @worachate001 could you share your location name to testing?

Bangkok, London, Moscow, with different Time Zone. They show the same results (and different magnitudes for Mars)!

**Comment by worachate001** [comment]
May I suggest an improvement to make the result much more accurate?

```
bool AstroCalcDialog::findPreciseQuadrature(QPair<double, double>* out, PlanetP object1, StelObjectP object2, double JD, double stopJD, double step)
{
	if (out == Q_NULLPTR)
		return false;

	step = -step / 288.;

	while (true)
	{
		JD += step;
		double dist = findDistance(JD, object1, object2, PhenomenaTypeIndex::Conjunction);
		double distNext = findDistance(JD + 1., object1, object2, PhenomenaTypeIndex::Conjunction);
		if (((dist-M_PI_2) < 0. && (distNext-M_PI_2) > 0.) || ((dist-M_PI_2) > 0. && (distNext-M_PI_2) < 0.))
		{
			double quadratureJD = JD+(qAbs(dist-M_PI_2)/qAbs(dist-distNext));
			out->first = quadratureJD;
			out->second = findDistance(quadratureJD, object1, object2, PhenomenaTypeIndex::Conjunction);
			if (object1->getJ2000EquatorialPos(core).longitude()>object2->getJ2000EquatorialPos(core).longitude())
				out->second *= -1.0; // let's use negative value for eastern quadratures
			return true;
		}
		else
			return false;

		if (JD > stopJD)
			return false;
	}
}
```
This will make the results accurate to a few minutes (and the planets are very close to the line of quadrature that displayed in the sky).

**Comment by alex-w** [comment]
@worachate001 thank you very much for very good patch!

**Comment by worachate001** [comment]
There must be a bug. Some western quadrature are shown as eastern and vice versa.

Something maybe wrong on this line.
`
if (object1->getJ2000EquatorialPos(core).longitude()>object2->getJ2000EquatorialPos(core).longitude())
`

**Comment by worachate001** [comment]
The issue still exist. If you search for the events for Jupiter between 2001-2010, you can see that there are many consecutive eastern quadratures. The problem happens because object->getJ2000EquatorialPos(core).longitude() can be negative. I would like to suggest changing the return of this function:

```
bool AstroCalcDialog::isSecondObjectRight(double JD, PlanetP object1, StelObjectP object2)
{
	core->setJD(JD);
	core->update(0);
	const double angle1 = object1->getJ2000EquatorialPos(core).longitude() * M_180_PI;
	const double angle2 = object2->getJ2000EquatorialPos(core).longitude() * M_180_PI;
	return (StelUtils::fmodpos(angle2-angle1, 360.)<180.);
}
```

**Comment by alex-w** [comment]
Good solution, thanks!

**Comment by axd1967** [comment]
The Astrocalc tool contains a conflict between "maximum separation" (currently limited to 20) and quadrature computation (separation 90!).

Try to avoid such artificial or obscure limitations. Why 20? Why not 21? Or 19?

Either remove the limit, or explain clearly why such a limit exists. Eg is it a fundamental issue, e.g. is it impossible to have separations of more than 20 degrees? (of course not.) is it a matter of accuracy that Stellarium cannot provide beyond 20 degrees? 

Another example: was the limit chosen because computation time might explode? If computation time might be a problem, leave the decision to the user rather than tie his/her hands to their back; e.g. the user will have to accept that larger separations mean more calculations. And some users might have access to powerful computers, they will not be happy.

Remember that the user might come up with interesting uses of apparently odd values.

**Comment by axd1967** [comment]
A quadrature line exists for every planet. There is no prior reason to not drawing the quadrature line when located on e.g. Mars. Outer planets will still go through quadrature points as the angle measure tool demonstrates.

(Similar for O/C line.)

image

**Comment by alex-w** [comment]
> A quadrature line exists for every planet. There is no prior reason to not drawing the quadrature line when located on e.g. Mars. Outer planets will still go through quadrature points as the angle measure tool demonstrates.

Please define the ecliptical coordinate systems for all planets of Solar system for current epoch.

**Comment by alex-w** [comment]
> The Astrocalc tool contains a conflict between "maximum separation" (currently limited to 20) and quadrature computation (separation 90!).

The "conflict" is not exist and it was not exist since introducing the Phenomena tool. The maximum separation parameter is used for filtering the conjunctions  and oppositions as tolerance for the angular distance from 0/180 degrees. This is obvious. 

> Try to avoid such artificial or obscure limitations. Why 20? Why not 21? Or 19?

Because conjunctions over 10 degrees is no meaning in astronomy. 20 degrees is selected as a twice of the good limit for conjunctions.

> Either remove the limit, or explain clearly why such a limit exists. Eg is it a fundamental issue, e.g. is it impossible to have separations of more than 20 degrees? (of course not.) is it a matter of accuracy that Stellarium cannot provide beyond 20 degrees?

Stellarium is an astronomical software! Conjunctions with 30, 45 and 60 degrees are used in astrological tasks only. If you want to get an astrological horoscope then you need use astrological software.

> Another example: was the limit chosen because computation time might explode? If computation time might be a problem, leave the decision to the user rather than tie his/her hands to their back; e.g. the user will have to accept that larger separations mean more calculations. And some users might have access to powerful computers, they will not be happy.
> 
> Remember that the user might come up with interesting uses of apparently odd values.

They should use some other odd software for it, which allow to get a strange, unrealistic or unmeaning data.

**Comment by axd1967** [comment]
> > A quadrature line exists for every planet. There is no prior reason to not drawing the quadrature line when located on e.g. Mars. Outer planets will still go through quadrature points as the angle measure tool demonstrates.
> 
> Please define the ecliptical coordinate systems for all planets of Solar system for current epoch.

I see - the plotting routines need a conversion between coordinate systems.

How about a curve that is 90 degrees away from the Sun or ASP? These are known.

The angle measure tool is able to compute angles, so an iterative process could be used based on the angle measurement routines to find the locus of all points at (eg) 90 degrees from ASP or Sun?

image


Now we need to find someone who wants to implement this ;-)

**Comment by axd1967** [comment]
> > The Astrocalc tool contains a conflict between "maximum separation" (currently limited to 20) and quadrature computation (separation 90!).
> 
> The "conflict" is not exist and it was not exist since introducing the Phenomena tool. The maximum separation parameter is used for filtering the conjunctions and oppositions as tolerance for the angular distance from 0/180 degrees. This is obvious.

Unfortunately, that's not that obvious. "**tolerance**" is a better term then, so that this confusion can be removed. Here is a suggestion:

image


> 
> > Try to avoid such artificial or obscure limitations. Why 20? Why not 21? Or 19?
> 
> Because conjunctions over 10 degrees is no meaning in astronomy. 20 degrees is selected as a twice of the good limit for conjunctions.

I understand perfectly; conjunction has a well-defined meaning. But if more than 10 degrees is meaningless (your words), why allow 20 then? What authority says that conjunctions above 20 degrees are meaningless? By loosening the artificial limit, Stellarium can help _both_ types of users: astronomers as well as "astrological nuts". Everybody happy, Stellarium gets a larger user base.

For example, a conjunction _tolerance_ of 180 degrees could allow to find moments when two planets could be visible at the same time, or when two planets would be opposite to each other. _And again, I'm not thinking of astrology, OK?_ So, imposing a limit should be very well thought through.

> 
> > Either remove the limit, or explain clearly why such a limit exists. Eg is it a fundamental issue, e.g. is it impossible to have separations of more than 20 degrees? (of course not.) is it a matter of accuracy that Stellarium cannot provide beyond 20 degrees?
> 
> Stellarium is an astronomical software! Conjunctions with 30, 45 and 60 degrees are used in astrological tasks only. If you want to get an astrological horoscope then you need use astrological software.

I'm not interested in astrological issues, alex-w. 

All I'm saying is that there is no need to artificially limit the user, unless there is some hard boundary (eg it doesn't make sense to allow a tolerance and/or separation of more than 180 degrees - or maybe even 360 under specific circumstances, eg if a sign is allowed to be involved). If someone wants to do astrological stuff with Stellarium, that's their right; as long as the calculations are mathematically/scientifically sound.

> 
> > Remember that the user might come up with interesting uses of apparently odd values.
> 
> They should use some other odd software for it, which allow to get a strange, unrealistic or unmeaning data.

Again, that's their problem. Stellarium devs don't know what can be done; their boundaries are mathematical/technical limits (avoiding divisions by zero, for example).

**Comment by alex-w** [comment]
> > Please define the ecliptical coordinate systems for all planets of Solar system for current epoch.
> 
> I see - the plotting routines need a conversion between coordinate systems.

Any plotting required coordinate system, always! This is fundamental property of our world.

> How about a curve that is 90 degrees away from the Sun or ASP? These are known.

No, this is unknown if you unknown coordinate system.

> The angle measure tool is able to compute angles, so an iterative process could be used based on the angle measurement routines to find the locus of all points at (eg) 90 degrees from ASP or Sun?

Angle measure tool uses equtarial and horizontal coordinate systems.

**Comment by gzotti** [comment]
Our conjunction finder attempts to find minimum distance between two moving objects at times when they come "close" in the sky to provide some sort of interesting view. A true conjunction in the classic sense requires equality in ecliptic longitude.  Quadrature originally meant 90° of difference in ecliptic longitude. The other (30/40/45/60/120°) ecliptic distance differences are of astrological significance only. You (yes, you!) can provide a plugin for them.  We don't care too much about enlarging our user base to specifically include astrologers, although I personally am interested in historical matters. I am pretty sure there is software out there which better caters the specific needs of astrologers, so we don't have to replicate everything. 

In this minimum finder, we alleviate this criterion and allow some wider separation in case these planets do not reach true conjunction but still provide an interesting view e.g. a pair of planets in binoculars.  A minimum distance of e.g. 50° between Venus and Saturn cannot be said to be an eye-catching and interesting view, so such minimum should not be listed in a list of "conjunctions". And discussing a minimum distance of 180° (i.e. opposition) is plainly nonsense in the context of "conjunction", sorry. We can discuss making the UI more complicated with the addition of yet another limitation switch to let the user limit results to 1...20° (oh yes, why limit to 20? Because this is what we, designers of Stellarium, regard potentially useful!)

An (unpartitioned) quadrature circle could indeed be plotted around the ASP, complementing the OC line, if anybody finds such feature useful. Does it help in any way, e.g. asteroid hunting?

**Comment by alex-w** [comment]
> Unfortunately, that's not that obvious. "**tolerance**" is a better term then, so that this confusion can be removed. Here is a suggestion:

Please open any astronomical dictionary and check meaning of "tolerance" and "angular separation" words.

> I understand perfectly; conjunction has a well-defined meaning. But if more than 10 degrees is meaningless (your words), why allow 20 then? What authority says that conjunctions above 20 degrees are meaningless? By loosening the artificial limit, Stellarium can help _both_ types of users: astronomers as well as "astrological nuts". Everybody happy, Stellarium gets a larger user base.

Probably you don't understand...

> For example, a conjunction _tolerance_ of 180 degrees could allow to find moments when two planets could be visible at the same time, or when two planets would be opposite to each other. _And again, I'm not thinking of astrology, OK?_ So, imposing a limit should be very well thought through.

I see you don't understand what meaning term "conjunction" in astronomy :( When maximum separation is setting to 10 degrees it's meaning that the tool should find all conjunctions where angular separation will less 10 degrees. Not equals exactly 10 degrees! If someone want to find and show only closes conjunctions then he will use this filter and as an expected he'll get all closes conjunctions. You're proposes remove this helpful feature.

**Comment by alex-w** [comment]
> An (unpartitioned) quadrature circle could indeed be plotted around the ASP, complementing the OC line, if anybody finds such feature useful. Does it help in any way, e.g. asteroid hunting?

But I've add rendering a quadrature circle in this PR ;)

**Comment by alex-w** [comment]
@gzotti please review the pull request

File: src/core/modules/GridLinesMgr.cpp [hunk 1]
```diff
@@ -1862,39 +1885,40 @@ void GridLinesMgr::init()
 	connect(&app, SIGNAL(screenFontSizeChanged(int)), this, SLOT(setFontSizeFromApp(int)));
 	
 	QString displayGroup = N_("Display Options");
-	addAction("actionShow_Gridlines",			displayGroup, N_("Grids and lines"), "gridlinesDisplayed");
+	addAction("actionShow_Gridlines",		displayGroup, N_("Grids and lines"), "gridlinesDisplayed");
```
Comment (gzotti) [inline src/core/modules/GridLinesMgr.cpp:1888]: please adjust indentation or whitespace before "displayGroup" with spaces only. Tabs apparently always are different between Win/Linux/Github-webview.

File: src/core/modules/GridLinesMgr.cpp [hunk 2]
```diff
@@ -2885,6 +2912,33 @@ void GridLinesMgr::setColorLongitudeLine(const Vec3f& newColor)
...
+//! Set flag for displaying quadrature line
+void GridLinesMgr::setFlagQuadratureLine(const bool displayed)
+{
+	if(displayed != quadratureLine->isDisplayed())
+	{
+		quadratureLine->setDisplayed(displayed);
+		emit quadratureLineDisplayedChanged(displayed);
+	}
+}
+//! Get flag for displaying quadratude line
```
Comment (gzotti) [inline src/core/modules/GridLinesMgr.cpp:2924]: typo in doc line: quadrature

File: src/gui/AstroCalcDialog.hpp [hunk 3]
```diff
@@ -540,6 +542,9 @@ private slots:
 	//! Calculation greatest elongations
 	QMap<double, double> findGreatestElongationApproach(PlanetP& object1, StelObjectP& object2, double startJD, double stopJD);
 	bool findPreciseGreatestElongation(QPair<double, double>* out, PlanetP object1, StelObjectP object2, double JD, double stopJD, double step);
+	//! Calculation guadratures
```
Comment (gzotti) [inline src/gui/AstroCalcDialog.hpp:545]: typo

File: src/gui/ViewDialog.cpp [hunk 4]
```diff
@@ -345,16 +345,17 @@ void ViewDialog::createDialogContent()
 	connect(ui->pushButtonAddRemoveLandscapes, SIGNAL(clicked()), this, SLOT(showAddRemoveLandscapesDialog()));
 
 	// Grid and lines
-	connectGroupBox(ui->celestialSphereGroupBox,				"actionShow_Gridlines");
+	connectGroupBox(ui->celestialSphereGroupBox,			"actionShow_Gridlines");
```
Comment (gzotti) [inline src/gui/ViewDialog.cpp:348]: again, for readability, spaces, not tabs, should be used to align the second column.

File: src/gui/viewDialog.ui [hunk 5]
```diff
@@ -4207,177 +4247,154 @@
...
                  </property>
                 </widget>
                </item>
-               <item row="11" column="6">
-                <widget class="QToolButton" name="colorInvariablePlane">
+               <item row="12" column="7">
+                <widget class="QCheckBox" name="showQuadratureLineCheckBox">
                  <property name="toolTip">
-                  <string>Color of the invariable plane of the Solar system</string>
+                  <string>The line of quadratures</string>
```
Comment (gzotti) [inline src/gui/viewDialog.ui:4394]: Better: A circle 90° from the Sun marking quadrature.

File: src/gui/viewDialog.ui [hunk 6]
```diff
@@ -4207,177 +4247,154 @@
...
-               <item row="1" column="2">
-                <widget class="QToolButton" name="colorEquatorLine">
+               <item row="12" column="6">
+                <widget class="QToolButton" name="colorQuadratureLine">
                  <property name="toolTip">
-                  <string>Color of equator (of date)</string>
-                 </property>
-                 <property name="text">
-                  <string notr="true"/>
+                  <string>Color of the quadratures line</string>
```
Comment (gzotti) [inline src/gui/viewDialog.ui:4307]: Better: Color of the quadrature circle

**Comment by gzotti** [comment]
I will try my luck making the Quadrature circle work in non-earth locations.

**Comment by gzotti** [comment]
Still struggling with the circle drawn differently. But later tonight, sure!

**Comment by axd1967** [comment]
my main interests are in observational astronomy. so - once more - please stop pushing me in the astrological corner, because I'm not really interested in astrological aspects. making such suggestions are unnecessary and only fueling interpersonal conflicts, nobody wants that.

I'm quite aware of the definition of conjunction.

all I'm doing is trying to expand Stellarium in a way that remains scientifically, mathematically as well as logically acceptable; the Stellarium team is free to ignore such suggestions.

so, as explained [here](https://github.com/Stellarium/stellarium/issues/789#issuecomment-1179511488), the quadrature line gives a quick means to find the largest phase angle of outer objects such e.g. Mars or Saturn and therefore get the largest shadows; useful information for astrophotographers.

thank you for the contribution.

**Comment by gzotti** [comment]
What happened to the "color of earth umbra" button? 
Regarding button ordering, it may be better to move QC to below Antisolar Point.

**Comment by gzotti** [comment]
I tried a quadrature circle for non-earth locations. However, sometimes it did not cross the ecliptic poles, no idea why. So, this earth-only version remains for now. It's too rare that I want to photograph Saturn in quadrature from Mars that it would really bother me. 

When the previous comments are clarified it's good for me.

**Comment by alex-w** [comment]
> What happened to the "color of earth umbra" button?

Fixed.

> Regarding button ordering, it may be better to move QC to below Antisolar Point.

I think no, because Antisolar point lies on the OC longitude

**Comment by alex-w** [comment]
> When the previous comments are clarified it's good for me.

Shall I merge it or you want to do it yourself?

**Comment by gzotti** [comment]
> > Regarding button ordering, it may be better to move QC to below Antisolar Point.
> 
> I think no, because Antisolar point lies on the OC longitude

The current sequence on the right: 
* Antisolar point
* Earth Umbra
* Earth Penumbra
* Invariable plane of the Solar System
* Quadrature circle

What I dislike is esp. the invariable plane separating the QC from its "relatives", OC and ASP.
Given that Quadrature circle is centered on Sun or ASP and somewhat complements OC line and ASP, I would prefer to see 
* Antisolar point
* Quadrature circle
* Earth Umbra
* Earth Penumbra
* Invariable plane of the Solar System

or even start with the QC next to the OC line:

* Quadrature circle
* Antisolar point
* Earth Umbra
* Earth Penumbra
* Invariable plane of the Solar System

Then the QC center (=ASP) and umbra circles are in clear grouping, and the invariable plane does not "mentally" interfere.

**Comment by axd1967** [comment]
> I tried a quadrature circle for non-earth locations. However, sometimes it did not cross the ecliptic poles, no idea why. So, this earth-only version remains for now. It's too rare that I want to photograph Saturn in quadrature from Mars that it would really bother me.

I hope this is [hyperbolic](https://github.com/Stellarium/stellarium/issues?q=[hyperbolic]); nobody ever hoped to photograph Saturn from Mars (at least... for now). Yet a quadrature line makes perfect sense on any SSB.

**Comment by alex-w** [comment]
> I hope this is a hyperbole; nobody ever hoped to photograph Saturn from Mars (at least... for now). Yet a quadrature line makes perfect sense on any SSB.

See #1518

**Pull request merged**